File: src/sim/canalPeriodCosts.ts
import type { GameState, ModuleId } from "./game";

export type CanalPeriodCosts = {
  venueBase: number;
  utilitiesAndCleaning: number;
  facilities: number;
  total: number;
};

const FIXED_VENUE_BASE = 300;
const FIXED_UTILITIES_AND_CLEANING = 212;

const FACILITY_PERIOD_COST: Partial<Record<ModuleId, number>> = {
  shower: 20,
  "cold-plunge": 75,
  "aufguss-yard": 100,
  program: 220,
};

function operationalForPeriod(snapshot: GameState, id: ModuleId) {
  if (!snapshot.built.includes(id)) return false;
  if (snapshot.repairTask?.moduleId === id) return false;
  if (id === "program" || id === "shower" || id === "cold-plunge") {
    return (snapshot.condition[id] ?? 100) > 0;
  }
  return true;
}

/**
 * Fixed obligations are deliberately not spread across guest blocks. They are period-boundary
 * costs: the venue exists for the week even if no guest crosses the door. Variable hour/session
 * costs belong to operating blocks and are modelled separately.
 */
export function calculateCanalPeriodCosts(snapshot: GameState): CanalPeriodCosts {
  const facilities = Object.entries(FACILITY_PERIOD_COST).reduce((total, [id, cost]) => {
    return total + (operationalForPeriod(snapshot, id as ModuleId) ? (cost ?? 0) : 0);
  }, 0);
  const total = FIXED_VENUE_BASE + FIXED_UTILITIES_AND_CLEANING + facilities;
  return {
    venueBase: FIXED_VENUE_BASE,
    utilitiesAndCleaning: FIXED_UTILITIES_AND_CLEANING,
    facilities,
    total,
  };
}

File: native/SaunaCore/Sources/SaunaCore/CanalPeriodCosts.swift
import Foundation

public struct CanalPeriodCostSummary: Equatable, Sendable {
    public let venueBase: Double
    public let utilitiesAndCleaning: Double
    public let facilities: Double

    public var total: Double { venueBase + utilitiesAndCleaning + facilities }
}

public enum CanalPeriodCosts {
    public static func calculate(built: Set<String>) -> CanalPeriodCostSummary {
        var facilities = 0.0
        if built.contains("shower") { facilities += 20 }
        if built.contains("cold-plunge") { facilities += 75 }
        if built.contains("aufguss-yard") { facilities += 190 }
        if built.contains("program") { facilities += 290 }
        return CanalPeriodCostSummary(
            venueBase: 300,
            utilitiesAndCleaning: 212,
            facilities: facilities
        )
    }
}
